core/test/network-records-to-devtools-log.js

```diff
@@ -225,28 +242,87 @@ function addRedirectResponseIfNeeded(networkRecords, record) {
  * @return {LH.DevtoolsLog}
  */
 function networkRecordsToDevtoolsLog(networkRecords, options = {}) {
+  // Clone test network records objects before potential modifications.
+  networkRecords = networkRecords.map(record => {
+    if (record.constructor === Object) {
+      record = {...record};
+      if (record.timing) record.timing = {...record.timing};
+    }
+
+    return record;
+  });
+
   const devtoolsLog = [];
-  networkRecords.forEach((networkRecord, index) => {
-    networkRecord = addRedirectResponseIfNeeded(networkRecords, networkRecord);
-    devtoolsLog.push(getRequestWillBeSentEvent(networkRecord, index));
+  networkRecords.forEach((record, index) => {
+    // If we're operating on network record raw objects (not NetworkRequest instances),
+    // then we're operating on test data that may need to be massaged a bit.
+    if (record.constructor === Object) {
+      // Temporary code while we transition away from startTime and endTime.
+      // This allows us to defer slightly changes to test files.
+      // TODO: remove after timing refactor is done
+      // See https://github.com/GoogleChrome/lighthouse/pull/14311
+      if (record.startTime !== undefined) {
+        record.networkRequestTime = record.startTime;
+        // Old tests never distinguished between these two.
+        record.rendererStartTime = record.startTime;
+      }
+      if (record.endTime !== undefined) {
+        record.networkEndTime = record.endTime;
+      }
+      if (record.responseReceivedTime !== undefined) {
+        record.responseHeadersEndTime = record.responseReceivedTime;
+      }
+
+      // TODO: good?
+      // if (record.rendererStartTime === undefined ^ record.networkEndTime === undefined) {
+      //   // eslint-disable-next-line max-len
+      //   throw new Error('expected test record to define rendererStartTime and networkEndTime together, but never just one');
+      // }
+
+      // Set timing.requestTime and timing.receiveHeadersEnd to be values that
+      // NetworkRequest will pull from for networkRequestTime and responseHeadersEndTime,
+      // so this roundtrips correctly. Unless, of course, timing values are explicitly set
+      // already.
+
+      // If record.timing explicitly does not having a timing object, only create one
+      // if we absolutely need it. See _recomputeTimesWithResourceTiming.
+      const netReqTime = record.networkRequestTime;
+      const willNeedTimingObject =
+        (netReqTime !== undefined && netReqTime !== record.rendererStartTime) ||
+        (record.responseHeadersEndTime !== undefined);
+      if (willNeedTimingObject) record.timing = record.timing || {};
```

Comment:
Ah, I recall why I did it this way. See the test failure for `yarn mocha gatherers/inspector-issues`:

```
      "sessionId": undefined,
      "statusCode": -1,
-     "timing": undefined,
+     "timing": Object {
+       "receiveHeadersEnd": 0,
+       "requestTime": -0.001,
+     },
      "transferSize": 0,
      "url": "https://example.com",
    },
  ]
      at networkRecordsToDevtoolsLog (file:///Users/cjamcl/src/lighthouse/core/test/network-records-to-devtools-log.js:295:40)
      at Context.<anonymous> (file:///Users/cjamcl/src/lighthouse/core/test/gather/gatherers/inspector-issues-test.js:329:19)
      at processImmediate (node:internal/timers:466:21)
```

Since timing object presence is now being automatically forced, the roundtrip assertion trips up unless we do the modification on the (cloned) test record.

So uh.... should I put it back?

Reply:
Part of it is `inspector-issues-test` is weird: it's `Object.assign`ing to ` real `NetworkRequest`, then passing it through `networkRecordsToDevtoolsLog` (it could have just used the partial passed through `networkRecordsToDevtoolsLog`, but it looks like those lines came a long time after the initial `mockRequest` method was written).

But even fixing that, there's a deeper problem and I don't think we can get away with optionally adding `timing` anymore. If you add

```diff
diff --git a/core/test/network-records-to-devtools-log-test.js b/core/test/network-records-to-devtools-log-test.js
index d114d483b..dd176e057 100644
--- a/core/test/network-records-to-devtools-log-test.js
+++ b/core/test/network-records-to-devtools-log-test.js
@@ -20,6 +20,11 @@ describe('networkRecordsToDevtoolsLog', () => {
       {method: 'Network.dataReceived'},
       {method: 'Network.loadingFinished'},
     ]);
+
+    const roundTripRecords = NetworkRecorder.recordsFromLogs(log);
+    const logNo2 = networkRecordsToDevtoolsLog(roundTripRecords);
+    const roundTripRecordsNo2 = NetworkRecorder.recordsFromLogs(logNo2);
+    expect(roundTripRecordsNo2).toMatchObject(roundTripRecords);
   });
 
   it('should set resource and transfer sizes', () => {
```
and run `yarn mocha network-records-to-devtools-log` this fails on this branch but passes on `main`. The first round-trip check (inside `networkRecordsToDevtoolsLog`) passes, but running the output through it again it fails on the same `timing` `undefined` vs `Object` part (the first time through it doesn't add a `timing` object, but the second time it does)